src/storage/cache_proxy/cluster_manager.rs

```diff
@@ -0,0 +1,1096 @@
+use futures::StreamExt;
+use tracing::{debug, warn};
+
+use crate::async_fuse::memfs::kv_engine::etcd_impl::Session;
+use crate::async_fuse::memfs::kv_engine::{KVEngine, SetOption};
+use crate::async_fuse::memfs::kv_engine::{KVEngineType, KeyType, ValueType};
+use crate::common::error::{DatenLordError, DatenLordResult};
+
+use super::node::{MasterNodeInfo, Node, NodeStatus};
+use super::ring::Ring;
+
+/// The timeout for current node's session
+const SESSION_TIMEOUT_SEC: u64 = 10;
+
+/// This struct is used to interact with etcd server and manager the cluster.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub struct ClusterManager {
+    /// inner struct
+    inner: Arc<ClusterManagerInner>,
+}
+
+#[allow(dead_code)]
+impl ClusterManager {
+    /// Create new cluster manager
+    pub fn new(kv_engine: Arc<KVEngineType>, node: Node) -> Self {
+        let inner = Arc::new(ClusterManagerInner::new(kv_engine, node));
+        Self { inner }
+    }
+
+    /// Get current node
+    #[must_use]
+    pub fn get_node(&self) -> Node {
+        self.inner.get_node()
+    }
+
+    /// Get current hashring
+    pub async fn get_ring(&self) -> DatenLordResult<Ring<Node>> {
+        self.inner.get_ring(false).await
+    }
+
+    /// Get all nodes
+    pub async fn get_nodes(&self) -> DatenLordResult<Vec<Node>> {
+        self.inner.get_nodes(false).await
+    }
+
+    ///  Stop the cluster manager
+    pub fn stop(&self) {
+        // In this step, we just to clean session and force down the cluster manager
+        self.inner.clean_sessions();
+    }
+
+    /// Run the cluster manager as state machine
+    ///
+    /// We need to perpare current node info, current node list and generate hash ring info
+    pub async fn run(&self) -> DatenLordResult<()> {
+        loop {
+            // 0. Prepare for the session
+            self.inner.prepare().await?;
+
+            // 1. Init cluster manager
+            debug!(
+                "Cluster manager start to run in node: {:?}",
+                self.inner.get_node().ip()
+            );
+            let mut current_node_info = self.inner.get_node();
+            // Next step is to register the node
+            current_node_info.set_status(NodeStatus::Registering);
+            self.inner.update_node(current_node_info);
+            self.inner.update_node_in_cluster().await?;
+
+            let mut current_node_info = self.inner.get_node();
+            // 2. Register node to etcd
+            debug!(
+                "Current node {:?} status: {:?}",
+                current_node_info.ip(),
+                current_node_info.status()
+            );
+            while self.inner.register().await.is_err() {
+                warn!("Failed to register node, retry in 5s");
+                // Try to register node to etcd
+                tokio::time::sleep(Duration::from_secs(5)).await;
+            }
+            // Update node status to Registering
+            current_node_info.set_status(NodeStatus::Slave);
```

Comment:
"set_status Slave" should be put after do_compaign()

Reply:
Thanks, current node status is modified to `Slave` as default value, this line will be removed.